### PR TITLE
Improve logging and exception handling in run_docker_test

### PR DESCRIPTION
**Pull request by aludvik**
Signed-off-by: Adam Ludvik <ludvik@bitwise.io>

### CODE REVIEW

**Comment by nick-drozd** [comment]
`ERROR:__main__:Test timed out.
ERROR:__main__:Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'up', '--abort-on-container-exit']' timed out after 299.99972581863403 seconds
Traceback (most recent call last):
  File "/usr/lib/python3.5/subprocess.py", line 695, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1072, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1741, in _communicate
    self.wait(timeout=self._remaining_time(endtime))
  File "/usr/lib/python3.5/subprocess.py", line 1650, in wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'up', '--abort-on-container-exit']' timed out after 299.999628046634 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/project/sawtooth-core/bin/run_docker_test", line 92, in main
    subprocess.run(compose_up, check=True, timeout=timer.remaining())
  File "/usr/lib/python3.5/subprocess.py", line 700, in run
    stderr=stderr)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'up', '--abort-on-container-exit']' timed out after 299.99972581863403 seconds
INFO:__main__:Shutting down with: ['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']
ERROR:__main__:Shutting down the test timed out.
ERROR:__main__:Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after 0 seconds
Traceback (most recent call last):
  File "/usr/lib/python3.5/subprocess.py", line 695, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1072, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1741, in _communicate
    self.wait(timeout=self._remaining_time(endtime))
  File "/usr/lib/python3.5/subprocess.py", line 1650, in wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after -0.0004865409999865733 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/project/sawtooth-core/bin/run_docker_test", line 140, in main
    subprocess.run(compose_down, check=True, timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 700, in run
    stderr=stderr)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after 0 seconds
ERROR:__main__:Shutting down the test timed out.
ERROR:__main__:Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after 30 seconds
Traceback (most recent call last):
  File "/usr/lib/python3.5/subprocess.py", line 695, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1072, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1741, in _communicate
    self.wait(timeout=self._remaining_time(endtime))
  File "/usr/lib/python3.5/subprocess.py", line 1650, in wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after 29.99992548299997 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/project/sawtooth-core/bin/run_docker_test", line 140, in main
    subprocess.run(compose_down, check=True, timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 700, in run
    stderr=stderr)
subprocess.TimeoutExpired: Command '['docker-compose', '-p', 'latest', '-f', 'poet-smoke.yaml', 'down', '--remove-orphans']' timed out after 30 seconds
CRITICAL:__main__:There are residual containers on the host that need to be cleaned up! Do `docker ps -a` and `docker newtork list` to see what was left behind!
`

**Comment by aludvik** [comment]
@nick-drozd That is telling me poet-smoke is timing out, then run_docker_test tries to clean it up with compose-down, which times out after 30 seconds. What is your comment/requested change?

**Comment by nick-drozd** [comment]
Is that the expected / desired behavior? If so, then there's no problem, and I approve. I ran it again after that one and it worked fine.

**Comment by boydjohnson** [comment]
My thought would be that when the subprocess.TimeoutExpired is caught, to set the timeout to 30s or something. TimeoutExpired does seem the most common of the errors, and when that one happens the cleanup isn't being given any time, the first time through the loop.

File: bin/run_docker_test [hunk 1]
```diff
@@ -132,9 +132,11 @@ def main():
 
         LOGGER.info("Shutting down with: %s", str(compose_down))
         shutdown_success = False
-        timeout = timer.remaining()
         for i in range(2):
             if shutdown_success == False:
+
+                # Always give compose down at least 60 seconds to cleanup
+                timeout = max(timer.remaining(), 60)
```
Comment (peterschwarz) [inline bin/run_docker_test]: Make this 60 a constant (suggest MAX_TIMEOUT_SEC), and use it below as well.

File: bin/run_docker_test [hunk 2]
```diff
@@ -60,63 +76,102 @@ def main():
...
+        else: # cleaning
+            exit_status = 0
+
+        LOGGER.info("Shutting down with: %s", str(compose_down))
+        shutdown_success = False
+        for i in range(2):
+            if shutdown_success == False:
+
+                # Always give compose down at least 60 seconds to cleanup
+                timeout = max(timer.remaining(), 60)
```
Comment (peterschwarz) [inline bin/run_docker_test]: Make this 60 a constant (suggest `MAX_TIMEOUT_SEC`), and use it below as well.

**Comment by aludvik** [comment]
When the timeout is caught, subprocess kills whatever the process is. I have a loop around the compose down command to try compose down again with a 60 second timeout if there was an error.

I also added another commit which checks for existing containers that would be reused and adds a --clean flag. If there are any, the test exits and prints a recommendation for fixing the issue. The --clean flag just runs compose-down with the tests compose file.

File: bin/run_docker_test (hunk 2 above)
Comment (aludvik) [inline bin/run_docker_test]: Sure, sorry I was pushing/pulling to my Intel machine where I had PoET in a bad state to verify.

**Pull request merged**